README
ShivaVG
=============================

See AUTHORS for the list of contributors

ShivaVG is an open-source LGPL ANSI C implementation of the Khronos
Group OpenVG specification.

I.    BUILD
II.   TESTING
III.  IMPLEMENTATION STATUS
IV.   EXTENSIONS


I. BUILD
=============================

 * Prerequisites:

   OpenGL development libraries and headers should be installed.
   Othe than that, since it's ANSI C should compile with any modern
   C compiler. jpeglib needs to be installed for example programs
   that use images.

 * Compiling under UNIX systems:

   Read the INSTALL file for more detailed (though generic) directions.
   This library uses the standard ./configure ; make. The example
   programs are automatically compiled. However, compilation of each
   example program can be toggled by ./configure --with-example-xxx
   command where xxx denotes the name of the example. Run ./configure
   --help for a list of such options.

 * Compiling on Mac:

   No XCode project files provided yet. The easiest way is by just
   using gcc, in which case look under UNIX compiling section.

 * Compiling natively on Windows platform:

   Solution files are provided for Visual C++ version 7 and 8. For
   the example programs using images to compile, you will need the
   appropriate build of jpeglib to match your Visual C++ version.

 * Compiling in mingw / cygwin environment:

   Might work just as fine as any UNIX-flavored system, but hasn't
   been tested yet.


II. TESTING
=============================

There is no real testing suite yet. The example programs are there
just to play with what the implementation can currently do, but
can hardly provide any proper validation, since no reference images
are provided. Here is a description of each example program and
what features it highlights:

* test_vgu

  Constructs some path primitives using the VGU API.

* test_tiger

  The most simple performance test. It draws the well known svg
  tiger using just simple stroke and fill of solid colors. It
  consists of 240 paths.

* test_dash

  Shows different stroke dashing modes.

* test_linear

  A rectangle drawn using 3-color linear gradient fill paint

* test_radial

  A rectangle drawn using 3-color radial gradient fill paint

* test_interpolate

  Interpolates between two paths - an apple and a pear.

* test_image

	Images are drawn using VG_DRAW_IMAGE_MULTIPLY image mode to be
  multiplied with radial gradient fill paint.

* test_pattern

  An image is drawn in multiply mode with an image pattern fill
  paint.


III. IMPLEMENTATION STATUS
=============================

Khronos states in the OpenVG specification, that the contexts for all
their client APIs are expected to be created via the EGL API. Since
EGL to use with ShivaVG has not been implemented yet, there is a set
of extension functions provided for the task of creating, maintaining
and destroying the OpenVG context. (See next section EXTENSIONS for
details.)

What follows is a description of which functions or to what extent
a certain function has been implemented. When a function is marked
as PARTIALLY implemented, the TODO file or the comments in the code
itself would provide further clues.


* General:

vgGetError ............................ FULLY implemented
vgFlush ............................... FULLY implemented
vgFinish .............................. FULLY implemented

* Getters and setters:

vgSet ................................. FULLY implemented
vgSeti ................................ FULLY implemented
vgSetfv ............................... FULLY implemented
vgSetiv ............................... FULLY implemented
vgGetf ................................ FULLY implemented
vgGeti ................................ FULLY implemented
vgGetVectorSize ....................... FULLY implemented
vgGetfv ............................... FULLY implemented
vgGetiv ............................... FULLY implemented
vgSetParameterf ....................... FULLY implemented
vgSetParameteri ....................... FULLY implemented
vgSetParameterfv ...................... FULLY implemented
vgSetParameteriv ...................... FULLY implemented
vgGetParameterf ....................... FULLY implemented
vgGetParameteri ....................... FULLY implemented
vgGetParameterVectorSize............... FULLY implemented
vgGetParameterfv ...................... FULLY implemented
vgGetParameteriv ...................... FULLY implemented

* Matrix Manipulation:

vgLoadIdentity ........................ FULLY implemented
vgLoadMatrix .......................... FULLY implemented
vgGetMatrix ........................... FULLY implemented
vgMultMatrix .......................... FULLY implemented
vgTranslate ........................... FULLY implemented
vgScale ............................... FULLY implemented
vgShear ............................... FULLY implemented
vgRotate .............................. FULLY implemented

* Masking and Clearing:

vgMask ................................ NOT implemented
vgClear ............................... FULLY implemented

* Paths:

vgCreatePath .......................... FULLY implemented
vgClearPath ........................... FULLY implemented
vgDestroyPath ......................... FULLY implemented
vgRemovePathCapabilities .............. FULLY implemented
vgGetPathCapabilities ................. FULLY implemented
vgAppendPath .......................... FULLY implemented
vgAppendPathData ...................... FULLY implemented
vgModifyPathCoords .................... FULLY implemented
vgTransformPath ....................... FULLY implemented
vgInterpolatePath ..................... FULLY implemented
vgPathLength .......................... FULLY implemented
vgPointAlongPath ...................... NOT implemented
vgPathBounds .......................... FULLY implemented
vgPathTransformedBounds ............... FULLY implemented
vgDrawPath ............................ PARTIALLY implemented

* Paint:

vgCreatePaint ......................... FULLY implemented
vgDestroyPaint ........................ FULLY implemented
vgSetPaint ............................ FULLY implemented
vgGetPaint ............................ FULLY implemented
vgSetColor ............................ FULLY implemented
vgGetColor ............................ FULLY implemented
vgPaintPattern ........................ FULLY implemented

* Images:

vgCreateImage ......................... PARTIALLY implemented
vgDestroyImage ........................ FULLY implemented
vgClearImage .......................... FULLY implemented
vgImageSubData ........................ PARTIALLY implemented
vgGetImageSubData ..................... PARTIALLY implemented
vgChildImage .......................... NOT implemented
vgGetParent ........................... NOT implemented
vgCopyImage ........................... FULLY implemented
vgDrawImage ........................... PARTIALLY implemented
vgSetPixels ........................... FULLY implemented
vgWritePixels ......................... FULLY implemented
vgGetPixels ........................... FULLY implemented
vgReadPixels .......................... FULLY implemented
vgCopyPixels .......................... FULLY implemented

* Image Filters:

vgColorMatrix ......................... NOT implemented
vgConvolve ............................ NOT implemented
vgSeparableConvolve ................... NOT implemented
vgGaussianBlur ........................ NOT implemented
vgLookup .............................. NOT implemented
vgLookupSingle ........................ NOT implemented

* Hardware Queries:

vgHardwareQuery ....................... NOT implemented

* Renderer and Extension Information:

vgGetString ........................... FULLY implemented

* VGU

vguLine ............................... FULLY implemented
vguPolygon ............................ FULLY implemented
vguRect ............................... FULLY implemented
vguRoundRect .......................... FULLY implemented
vguEllipse ............................ FULLY implemented
vguArc ................................ FULLY implemented
vguComputeWarpQuadToSquare ............ NOT implemented
vguComputeWarpSquareToQuad ............ NOT implemented
vguComputeWarpQuadToQuad .............. NOT implemented


IV. EXTENSIONS
=============================

There are three extensions to the API that manipulate the OpenVG
context as a temporary replacement for EGL:

VGboolean vgCreateContextSH(VGint width, VGint height)

  Creates an OpenVG context on top of an existing OpenGL context
  that should have been manually initialized by the user of the
  library. Width and height specify the size of the rendering
  surface. No multi-threading support has been implemented yet.
  The context is created once per process.

void vgResizeSurfaceSH(VGint width, VGint height)

  Should be called whenever the size of the surface changes (e.g.
  the owner window of the OpenGL context is resized).

void vgDestroyContextSH()

  Destroys the OpenVG context associated with the calling process.


ChangeLog
21 Nov 2014 Dan Panzarella <alsoelp@gmail.com>
 
  * finish vgGetPathLength for all segment types

20 Nov 2014 Dan Panzarella <alsoelp@gmail.com>

  * calculate vgGetPathLength for all but cubic beziers and arcs

18 Nov 2014 Dan Panzarella <alsoelp@gmail.com>

  * straighten out build artifacts with gitignore

14 Nov 2014 Dan Panzarella <alsoelp@gmail.com>

  * shPaint.c: implement vgGetPaint

13 Nov 2014 Dan Panzarella <alsoelp@gmail.com>

  * shDefs.h: fix compiling on recent Arch Linux

04 Feb 2008 Ivan Leben <ivan.leben@gmail.com>

  * shImage.c: shCopyPixels uses memcpy when formats are
    equal.

04 Feb 2008 Ivan Leben <ivan.leben@gmail.com>

  * openvg.h: changed typedef of VGHandle to void*. Code
    compiles with no warnings on 64-bit systems now and
    it's much safer anyway.

03 Feb 2008 Ivan Leben <ivan.leben@gmail.com>

  * shContest.c: fixed a bug in SHContext destructor where
    resource deletion loop was iterating image array while
    freeing paint objects

03 Feb 2008 Ivan Leben <ivan.leben@gmail.com>

  * a patch by Vincenzo Pupillo fixes array constructor behavior
    when it runs out of memory plus makes its destructor a bit
    more robust

03 Feb 2008 Ivan Leben <ivan.leben@gmail.com>

  * autogen.sh: added --force to autoreconf call

03 Feb 2008 Ivan Leben <ivan.leben@gmail.com>

  * Removed automake/autoconf generated files. Added autogen.sh
    to generate them.

03 Feb 2008 Ivan Leben <ivan.leben@gmail.com>

  * configure.in: changed order of checks for glut library and
    glXGetProcAddress in -lGL so that latter doesn't fail when
    glut is not installed.

30 Jan 2008 Ivan Leben <ivan.leben@gmail.com>

  * shGeometry.c: shStrokePath() fixed to handle zero-length line
    segments gracefully

  * fixed the image color component ordering to be valid against the
    spec

  * shImage.c, shImage.h: completely overhauled to support multiple
    image formats. Everything but _PRE and _BW_1 now supported.

24 Jan 2008 Ivan Leben <ivan.leben@gmail.com>

  * Moved release-related info from ChangeLog to NEWS. ChangeLog will
    now include info regarding each commit to repository

  * Rearranged the path tesselletaion and user-to-surface transformation
    steps in the pipeline to produce higher curve resolution when path is
    scaled up

  * Using new pipeline design, transformed bound-box query can now return
    really tight bounds

  * Inserted empty/nop definitions for all non-implemented function so
    that linking against the library doesn't fail.

22 Jan 2008 Ivan Leben <ivan.leben@gmail.com>

  * Fixed a bug where vgGetError would always return VG_NO_ERROR


src/shVgu.c
/*
 * Copyright (c) 2007 Ivan Leben
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 * 
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 * 
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library in the file COPYING;
 * if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 *
 */

#define VG_API_EXPORT
#include "openvg.h"
#include "vgu.h"
#include "shDefs.h"
#include "shContext.h"
#include <math.h>

static VGUErrorCode shAppend(VGPath path, SHint commSize, const VGubyte *comm,
                             SHint dataSize, const VGfloat *data)
{
  VGErrorCode err = VG_NO_ERROR;
  VGPathDatatype type = vgGetParameterf(path, VG_PATH_DATATYPE);
  VGfloat scale = vgGetParameterf(path, VG_PATH_SCALE);
  VGfloat bias = vgGetParameterf(path, VG_PATH_BIAS);
  SH_ASSERT(dataSize <= 26);
  
  switch(type)
  {
  case VG_PATH_DATATYPE_S_8: {
      
      SHint8 data8[26]; int i;
      for (i=0; i<dataSize; ++i)
        data8[i] = (SHint8)SH_FLOOR((data[i] - bias) / scale + 0.5f);
      vgAppendPathData(path, commSize, comm, data8);
      
      break;}
  case VG_PATH_DATATYPE_S_16: {
      
      SHint16 data16[26]; int i;
      for (i=0; i<dataSize; ++i)
        data16[i] = (SHint16)SH_FLOOR((data[i] - bias) / scale + 0.5f);
      vgAppendPathData(path, commSize, comm, data16);
      
      break;}
  case VG_PATH_DATATYPE_S_32: {
      
      SHint32 data32[26]; int i;
      for (i=0; i<dataSize; ++i)
        data32[i] = (SHint32)SH_FLOOR((data[i] - bias) / scale + 0.5f);
      vgAppendPathData(path, commSize, comm, data32);
      
      break;}
  default: {
      
      VGfloat dataF[26]; int i;
      for (i=0; i<dataSize; ++i)
        dataF[i] = (data[i] - bias) / scale;
      vgAppendPathData(path, commSize, comm, dataF);
      
      break;}
  }
  
  err = vgGetError();
  if (err == VG_PATH_CAPABILITY_ERROR)
    return VGU_PATH_CAPABILITY_ERROR;
  else if (err == VG_BAD_HANDLE_ERROR)
    return VGU_BAD_HANDLE_ERROR;
  else if (err == VG_OUT_OF_MEMORY_ERROR)
    return VGU_OUT_OF_MEMORY_ERROR;
  
  return VGU_NO_ERROR;
}

VGU_API_CALL VGUErrorCode vguLine(VGPath path,
                                  VGfloat x0, VGfloat y0,
                                  VGfloat x1, VGfloat y1)
{
  VGUErrorCode err = VGU_NO_ERROR;
  const VGubyte comm[] = {VG_MOVE_TO_ABS, VG_LINE_TO_ABS};
  
  VGfloat data[4];
  data[0] = x0; data[1] = y0;
  data[2] = x1; data[3] = y1;
  
  err = shAppend(path, 2, comm, 4, data);
  return err;
}

VGU_API_CALL VGUErrorCode vguPolygon(VGPath path,
                                     const VGfloat * points, VGint count,
                                     VGboolean closed)
{
  VGint i;
  VGubyte *comm = NULL;
  VGUErrorCode err = VGU_NO_ERROR;
  
  if (points == NULL || count <= 0)
    return VGU_ILLEGAL_ARGUMENT_ERROR;
  /* TODO: check points array alignment */
  
  comm = (VGubyte*)malloc( (count+1) * sizeof(VGubyte) );
  if (comm == NULL) return VGU_OUT_OF_MEMORY_ERROR;
  
  comm[0] = VG_MOVE_TO_ABS;
  for (i=1; i<count; ++i)
    comm[i] = VG_LINE_TO_ABS;
  comm[count] = VG_CLOSE_PATH;
  
  if (closed) err = shAppend(path, count+1, comm, count*2, points);
  else        err = shAppend(path, count, comm, count*2, points);
  
  free(comm);
  return err;
}

VGU_API_CALL VGUErrorCode vguRect(VGPath path,
                                  VGfloat x, VGfloat y,
                                  VGfloat width, VGfloat height)
{
  VGUErrorCode err = VGU_NO_ERROR;
  
  VGubyte comm[5] = {
    VG_MOVE_TO_ABS, VG_HLINE_TO_REL,
    VG_VLINE_TO_REL, VG_HLINE_TO_REL,
    VG_CLOSE_PATH };
  
  VGfloat data[5];
  
  if (width <= 0 || height <= 0)
    return VGU_ILLEGAL_ARGUMENT_ERROR;
  
  data[0] =  x;     data[1] = y;
  data[2] =  width; data[3] = height;
  data[4] = -width;
  
  err = shAppend(path, 5, comm, 5, data);
  return err;
}

VGU_API_CALL VGUErrorCode vguRoundRect(VGPath path,
                                       VGfloat x, VGfloat y,
                                       VGfloat width, VGfloat height,
                                       VGfloat arcWidth, VGfloat arcHeight)
{
  VGUErrorCode err = VGU_NO_ERROR;
  
  VGubyte comm[10] = {
    VG_MOVE_TO_ABS,
    VG_HLINE_TO_REL, VG_SCCWARC_TO_REL,
    VG_VLINE_TO_REL, VG_SCCWARC_TO_REL,
    VG_HLINE_TO_REL, VG_SCCWARC_TO_REL,
    VG_VLINE_TO_REL, VG_SCCWARC_TO_REL,
    VG_CLOSE_PATH };
  
  VGfloat data[26];
  VGfloat rx, ry;
  
  if (width <= 0 || height <= 0)
    return VGU_ILLEGAL_ARGUMENT_ERROR;
  
  SH_CLAMP(arcWidth, 0.0f, width);
  SH_CLAMP(arcHeight, 0.0f, height);
  rx = arcWidth/2;
  ry = arcHeight/2;
  
  data[0]  =  x + rx; data[1] = y;
  
  data[2]  =  width - arcWidth;
  data[3]  =  rx; data[4]  =  ry; data[5]  = 0;
  data[6]  =  rx; data[7]  =  ry;
  
  data[8]  =  height - arcHeight;
  data[9]  =  rx; data[10] =  ry; data[11] = 0;
  data[12] = -rx; data[13] =  ry;
  
  data[14] = -(width - arcWidth);
  data[15] =  rx; data[16] =  ry; data[17] = 0;
  data[18] = -rx; data[19] = -ry;
  
  data[20] = -(height - arcHeight);
  data[21] =  rx; data[22] =  ry; data[23] = 0;
  data[24] =  rx; data[25] = -ry;
  
  err = shAppend(path, 10, comm, 26, data);
  return err;
}

VGU_API_CALL VGUErrorCode vguEllipse(VGPath path,
                                     VGfloat cx, VGfloat cy,
                                     VGfloat width, VGfloat height)
{
  VGUErrorCode err = VGU_NO_ERROR;
  
  const VGubyte comm[] = {
    VG_MOVE_TO_ABS, VG_SCCWARC_TO_REL,
    VG_SCCWARC_TO_REL, VG_CLOSE_PATH};
  
  VGfloat data[12];
  
  if (width <= 0 || height <= 0)
    return VGU_ILLEGAL_ARGUMENT_ERROR;
  
  data[0] = cx + width/2; data[1] = cy;
  
  data[2] = width/2; data[3] = height/2; data[4] = 0;
  data[5] = -width; data[6] = 0;
  
  data[7] = width/2; data[8] = height/2; data[9] = 0;
  data[10] = width; data[11] = 0;
  
  err = shAppend(path, 4, comm, 12, data);
  return err;
}

#include <stdio.h>

VGU_API_CALL VGUErrorCode vguArc(VGPath path,
                                 VGfloat x, VGfloat y,
                                 VGfloat width, VGfloat height,
                                 VGfloat startAngle, VGfloat angleExtent,
                                 VGUArcType arcType)
{
  VGUErrorCode err = VGU_NO_ERROR;
  
  VGubyte commStart[1] = {VG_MOVE_TO_ABS};
  VGfloat dataStart[2];
  
  VGubyte commArcCCW[1] = {VG_SCCWARC_TO_ABS};
  VGubyte commArcCW[1]  = {VG_SCWARC_TO_ABS};
  VGfloat dataArc[5];
  
  VGubyte commEndPie[2] = {VG_LINE_TO_ABS, VG_CLOSE_PATH};
  VGfloat dataEndPie[2];
  
  VGubyte commEndChord[1] = {VG_CLOSE_PATH};
  VGfloat dataEndChord[1] = {0.0f};
  
  VGfloat alast, a = 0.0f;
  VGfloat rx = width/2, ry = height/2;
  
  if (width <= 0 || height <= 0)
    return VGU_ILLEGAL_ARGUMENT_ERROR;
  
  if (arcType != VGU_ARC_OPEN &&
      arcType != VGU_ARC_CHORD &&
      arcType != VGU_ARC_PIE)
      return VGU_ILLEGAL_ARGUMENT_ERROR;
  
  startAngle = SH_DEG2RAD(startAngle);
  angleExtent = SH_DEG2RAD(angleExtent);
  alast = startAngle + angleExtent;
  
  dataStart[0] = x + SH_COS(startAngle) * rx;
  dataStart[1] = y + SH_SIN(startAngle) * ry;
  err = shAppend(path, 1, commStart, 2, dataStart);
  if (err != VGU_NO_ERROR) return err;
  
  dataArc[0] = rx;
  dataArc[1] = ry;
  dataArc[2] = 0.0f;
  
  if (angleExtent > 0) {
    
    a = startAngle + PI;
    while (a < alast) {
      dataArc[3] = x + SH_COS(a) * rx;
      dataArc[4] = y + SH_SIN(a) * ry;
      err = shAppend(path, 1, commArcCCW, 5, dataArc);
      if (err != VGU_NO_ERROR) return err;
      a += PI; }
    
    dataArc[3] = x + SH_COS(alast) * rx;
    dataArc[4] = y + SH_SIN(alast) * ry;
    err = shAppend(path, 1, commArcCCW, 5, dataArc);
    if (err != VGU_NO_ERROR) return err;
    
  }else{
    
    a = startAngle - PI;
    while (a > alast) {
      dataArc[3] = x + SH_COS(a) * rx;
      dataArc[4] = y + SH_SIN(a) * ry;
      err = shAppend(path, 1, commArcCW, 5, dataArc);
      if (err != VGU_NO_ERROR) return err;
      a -= PI; }
    
    dataArc[3] = x + SH_COS(alast) * rx;
    dataArc[4] = y + SH_SIN(alast) * ry;
    err = shAppend(path, 1, commArcCW, 5, dataArc);
    if (err != VGU_NO_ERROR) return err;
  }
  
  
  if (arcType == VGU_ARC_PIE) {
    dataEndPie[0] = x; dataEndPie[1] = y;
    err = shAppend(path, 2, commEndPie, 2, dataEndPie);
  }else if (arcType == VGU_ARC_CHORD) {
    err = shAppend(path, 1, commEndChord, 0, dataEndChord);
  }
  
  return err;
}
